### ISSUE TITLE
azure - auto-tag-date example code throwing stack trace

### ISSUE THREAD

**Issue by jmery** [cloud-custodian/cloud-custodian#4893]
```
python 3.7.4
c7n==0.8.44.3
c7n_azure==0.5.5
```
When attempting to run the sample policy for `auto-tag-date`, a stack trace is thrown.  It appears that `auto-tag-date` isn't a valid action type?

Sample code:
```
policies:
  - name: azure-auto-tag-created-date
    resource: azure.resourcegroup
    description: |
      Tag all existing resource groups with the 'CreatedDate' tag
    actions:
      - type: auto-tag-date
        tag: CreatedDate
        format: "%m-%d-%Y"
```
☝️ from https://cloudcustodian.io/docs/azure/resources/azure-common-actions.html#azure-common-actions-auto-tag-date

Stack trace:
```
Traceback (most recent call last):
  File "/Users/jmery/.pyenv/versions/resource_tagger/bin/custodian", line 11, in <module>
    load_entry_point('c7n==0.8.44.3', 'console_scripts', 'custodian')()
  File "/Users/jmery/.pyenv/versions/3.7.4/envs/resource_tagger/lib/python3.7/site-packages/c7n/cli.py", line 373, in main
    command(config)
  File "/Users/jmery/.pyenv/versions/3.7.4/envs/resource_tagger/lib/python3.7/site-packages/c7n/commands.py", line 65, in _load_policies
    collection = policy_load(options, fp, validate=validate, vars=vars)
  File "/Users/jmery/.pyenv/versions/3.7.4/envs/resource_tagger/lib/python3.7/site-packages/c7n/policy.py", line 63, in load
    errors[1], errors[0]))
c7n.exceptions.PolicyValidationError: Failed to validate policy azure-auto-tag-created-date 
 Error on policy:azure-auto-tag-created-date resource:azure.resourcegroup
{'type': 'auto-tag-date', 'tag': 'CreatedDate', 'format': '%m-%d-%Y'} is not valid under any of the given schemas

Failed validating 'anyOf' in schema[164]['allOf'][1]['properties']['actions']['items']:
    {'anyOf': [{'$ref': '#/definitions/actions/azure.notify'},
               {'$ref': '#/definitions/actions/azure.webhook'},
               {'$ref': '#/definitions/resources/azure.resourcegroup/actions/delete'},
               {'$ref': '#/definitions/actions/azure.logic-app'},
               {'$ref': '#/definitions/actions/azure.tag'},
               {'$ref': '#/definitions/actions/azure.untag'},
               {'$ref': '#/definitions/actions/azure.auto-tag-user'},
               {'$ref': '#/definitions/actions/azure.tag-trim'},
               {'$ref': '#/definitions/actions/azure.mark-for-op'},
               {'$ref': '#/definitions/actions/azure.lock'},
               {'enum': ['notify',
                         'webhook',
                         'delete',
                         'logic-app',
                         'tag',
                         'untag',
                         'auto-tag-user',
                         'tag-trim',
                         'mark-for-op',
                         'lock']}]}

On instance['actions'][0]:
    {'format': '%m-%d-%Y', 'tag': 'CreatedDate', 'type': 'auto-tag-date'}
```

**Comment by kapilt** [comment]
it looks you might be running an older version of the azure provider that doesn't have the auto-tag-date action.

try `pip install -U c7n_azure` and `pip install -U c7n` to update.

**Comment by jmery** [comment]
Looks like that did the trick.  Thanks.  On a related note ...

When deploying custodian for production, what are the best practices in pinning versions of custodian in relation to things like c7n_azure?

We currently pin both the custodian and tools based on the versions listed in the tagged release (e.g. https://github.com/cloud-custodian/cloud-custodian/commit/f6a98c6918c437033e635758302d72a25a8f09b8).

1. Do we even need to bother with pinning the tools versions (if it's pinned on the custodian side) and we're already pinning to the version of custodian?
2. Are there any concerns running different versions of the tools from what's specified?

Do we even need to bother with pinning the tools?

**Comment by kapilt** [comment]
production deploys are about getting consistent working sets across the packages for custodian, the provider in use, and the dependencies. generally we release lockstep for custodian core and across providers. generally the c7n to provider dependency consideration between the providers and the core is pretty loose, in that the provider will work across multiple versions of core, however its something we need to work on to appropriately record and update min core dependency version when there is a provider affecting core update. gcp and aws providers have stable underlying dependency graphs. however the azure provider has some additional concerns due to the underlying instability in the official python azure sdk (regular compatibility breaks). we've been working on a docker based install tooling as an alternative, and publish a tested docker image for every release and every commit to docker hub.

so right now for azure, I would recommend getting a released version tagged docker image, or alternatively pip freezing transitive requirements for a working environment post upgrade. if pulling from git, then running tests via tox to run tests to verify the dependency graph. @stefangordon @aluong what have you been recommending to azure users?

**Comment by kapilt** [comment]
I talked to @stefangordon about this on gitter, I think I'm going to take this as an action item/bug that we need to pin deps for c7n_azure on release to work around the instability of the azure sdk graph overtime, such that pypi based installs for c7n_azure are usable reliably.

**Comment by jmery** [comment]
Thanks @kapilt.  We're basically pip freezing via our install process (dogfooding one of our commercial products internally).  That works well enough for our purposes and it's good to see it inline with your recommendations.  Appreciate the help!

**Comment by kapilt** [comment]
cool, fwiw at least for the custodian azure provider package (c7n_azure), i think for the next release we're going to move to freezing the transitive dependency set on upload to pypi via https://github.com/cloud-custodian/cloud-custodian/pull/4901

**Comment by kapilt** [comment]
fwiw 0.8.45.1 was released to PyPI and c7n_azure 0.6.1 has frozen version requirements.
